bolt/include/bolt/Core/MCPlusBuilder.h

```diff
@@ -1412,9 +1413,15 @@ class MCPlusBuilder {
     return false;
   }
 
-  /// Modify a direct call instruction \p Inst with an indirect call taking
-  /// a destination from a memory location pointed by \p TargetLocation symbol.
-  virtual bool convertCallToIndirectCall(MCInst &Inst,
+  /// Modify a direct call instruction pointed by the iterator \p It, with an
+  /// indirect call taking a destination from a memory location pointed by \p
+  /// TargetLocation symbol. If additional instructions need to be prepended
+  /// before \p It, then the iterator must be updated to point to the indirect
+  /// call instruction.
+  ///
+  /// \return true on success
+  virtual bool convertCallToIndirectCall(BinaryBasicBlock &BB,
```

Comment:
Hi Paschalis, 

Thanks a lot for submitting this contribution to the project, I appreciate it!

One thing that might not be obvious in the code base is that the MCPlusBuilder layer (or the BOLT targets lib, for that matter) do not depend on BOLT IR classes, such as BinaryBasicBlock. They are limited to use MCInst and the class MCPlusBuilder::InstructionIterator whenever the API needs to communicate a walkable MCInst pointer to target-specific code.

If you add this dependency of bolt target libs on bolt IR classes, this will create a cyclic dependency that will likely break our shared lib build (if you build BOLT with enable_shared you might see this). I would suggest you taking a look at other methods in X86MCPlusBuilder to see how we implement this kind of optimization -- usually by moving code that depends on BB knowledge to the pass itself, instead of doing that logic in target libs.

Reply:
@paschalis-mpeis There is a polymorphic iterator `InstructionIterator`, which is defined in `MCPlusBuilder.h` and there is a pattern to pass every required instruction sequence to `MCPlusBuilder`'s methods by this iterator (but, as I see, it is used to pass a chain of instruction with no modifications of the iterator itself). Also there is the `InstructionListType` and, please have a look at the [`createLoadImmediate`](https://github.com/llvm/llvm-project/blob/main/bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp#L1453) for example, the generated in `MCPlusBuilder` instruction list should be returned by value.

Reply:
Thanks @rafaelauler  and @samolisov  for the suggestions. I think the way to go is using `InstructionListType`.

I will rename `convertCallToIndirectCall` to `createIndirectPltCall`, addressing my previous concern [here](https://github.com/llvm/llvm-project/pull/93584#discussion_r1617667657), but feel free to suggest a better name.
The function will be adjusted to return an `InstructionListType`, and anything that relates to Binary BB logic will be moved to the pass itself.

I think the closest existing example of this is with the `InlineMemcpy` pass and `createInlineMemcpy`.